### ISSUE TITLE
Build tensorflow 2.2.0 failed in '@upb//:upb' with the length argument in strncpy

### ISSUE THREAD

**Issue by jxy** [tensorflow/tensorflow#39467]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
  Linux Ubuntu 18.04.4 LTS
- TensorFlow installed from (source or binary):
  source
- TensorFlow version:
  2.2.0
- Python version:
  3.7.7
- Bazel version (if compiling from source):
  2.0.0
- GCC/Compiler version (if compiling from source):
  10.1.0

**Describe the problem**

```console
In file included from /usr/include/string.h:494,                                                                
                 from external/upb/upb/upb.h:16,                                                                                                                                                                                              
                 from external/upb/upb/upb.c:2:                                                                 
In function 'strncpy',                                                                                         
    inlined from 'upb_status_seterrmsg' at external/upb/upb/upb.c:40:3:                                         
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: error: '__builtin_strncpy' specified bound 127 equals destination size [-Werror=stringop-truncation]                                                             
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));                                                                                                                                                              
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                 
cc1: all warnings being treated as errors                                                                      
Target //tensorflow/tools/pip_package:build_pip_package failed to build                                         
```

**Provide the exact sequence of commands / steps that you executed before running into the problem**

Build with bazel following the documentation.

**Any other info / logs**

It's a trivial misuse of strncpy.  Here is a simple fix.

```patch
--- upb/upb/upb.c~      2020-05-12 14:33:15.046048687 -0500                                                                                                                                           
+++ upb/upb/upb.c       2020-05-12 14:33:30.281987830 -0500                                                                                                                                           
@@ -37,7 +37,7 @@                                                                                                                                                                                                                                             
 void upb_status_seterrmsg(upb_status *status, const char *msg) {                                                                                                                                                                                             
   if (!status) return;                                                                                                                                                                                                                                       
   status->ok = false;                                         
-  strncpy(status->msg, msg, sizeof(status->msg));             
+  strncpy(status->msg, msg, sizeof(status->msg)-1);           
   nullz(status);                                              
 }                                                             
                                                               
```

**Comment by jxy** [comment]
This issue is already fixed in upb upstream, so I posted it here.

**Comment by yongtang** [comment]
@jxy added a PR #39467 to update upb.

**Comment by yongtang** [comment]
@jxy  Update: Should be PR #39470.

**Comment by yongtang** [comment]
@jxy The fix is not straightforward, as grpc still uses repository_defs.bzl in upb which has been removed:

https://github.com/protocolbuffers/upb/blob/d7d72f00753767ae639c12f2a7618c56ccb2425a/bazel/repository_defs.bzl

Will need to patch grpc repo first. Then updates tf to get this fixed.

**Comment by mihaimaruseac** [comment]
Yes, this needs to wait for grpc to upgrade first.

**Comment by 0xA1B2** [comment]
When can we compile TF 2.2.0 on Clear Linux without this bug ?

**Comment by svenstaro** [comment]
The current state of this is that tensorflow sadly still doesn't work on gcc 10 which is especially annoying since even CUDA 11.1 now supports gcc 10. As a downstream packager, I'm not really sure what the proper way to address this problem would be but I'd appreciate it if a tensorflow team member could fix it.

**Comment by jpag94** [comment]
Second that!

**Comment by mihaimaruseac** [comment]
We don't update branches after release, except for security reasons. So to test if this is fixed, we should test on the r2.4 branch

**Comment by aviallon** [comment]
Of course, I was talking about the last branch here.

**Comment by yongtang** [comment]
The grpc upstream library has updated and includes the right upb library in https://github.com/grpc/grpc/commit/e2e6b1839e83b7373edda903fb17dd15a2cb996f (included in grpc 1.32.0/1.33.0/1.33.1/1.33.2), so updating tensorflow's grpc library in theory will resolve this issue.

Though it looks like there are quite a few other dependencies that might be interleaved with grpc dependency so the update is not exactly straightforward.

**Comment by yongtang** [comment]
Actually, a PR #44282 has already been opened to update grpc which will resolve this issue (if the PR is able to get merged).

**Comment by svenstaro** [comment]
This can be closed.

**Comment by mihaimaruseac** [comment]
Closing as requested

**Comment by mchinen** [comment]
I think this should be reopened due the rollback of #51923.

**Comment by ian-huu** [comment]
Any update?

**Comment by mihaimaruseac** [comment]
CC @learning-to-play : this is another issue blocked on protobuf/grpc bump

**Comment by thias** [comment]
I'm also running into this issue when trying to build TensorFlow Serving 2.14.1 on Red Hat Enterprise Linux 9 (only gcc 11 and later available). It's super frustrating to not be able to easily get a 2 character patch included in the build...

So far I have tried a bunch of different hacks, but to no avail. That upb not being pulled in directly by TensorFlow Serving workspace stuff makes it seem really tricky to me (and zero previous experience with Bazel doesn't help).

What workarounds might possibly work here?

**Comment by thias** [comment]
I managed to get a build to complete. The workaround (not a fix) is to add this to the bazel build command:
```
--copt=-Wno-stringop-truncation
```
It seems to make its way to the upb compilation and gets gcc to ignore the issue.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
